### ISSUE TITLE
NOT break all mod if one mod can't compiled

### ISSUE THREAD

**Issue by namkazt** [JHawkley/Harmony-Injector#2]
This being an issue from the beginning. Any ideas about how to deal with this?

**Comment by JHawkley** [comment]
Oh, that'd be an easy one: just don't compile all the mods into a single mod assembly; give each mod its own assembly.

It's trivial to just loop through each `ModInfo` in `ModManager.mods` and run the C# compiler on its source code individually; the Mono runtime doesn't much care if mods come in one assembly or twelve as long as they link properly.  If one mod fails to compile, it will only break those mods that were reliant on the mod that failed.

After that, it's just fixing type resolution in `ModManager`.  A `foreach` loop over the mod assemblies is all that would take, but you'd probably want to do it in reverse order, in order to respect mod load order.  A mod loaded later should override a type from a mod loaded earlier.

And then the game would be right back to working as it does now, just with mods spread across many assemblies.

**Comment by namkazt** [comment]
Yes I was trying to do this but the point is at the first load COQ alway compile all mods into 1 assembly

**Comment by JHawkley** [comment]
Yep.  And this is largely why my changes moved away from taking over compilation of mods from Qud.  It doesn't add much benefit; issues related to the loading of mods can't be solved by a mod, and making the attempt just introduces a potential point-of-failure.

...which is what broke the mod with the last game update.

I actually realized all this pretty late.  I had all these lovely error popups about compilation failures and the like in Harmony Injector...  Then realized few people would see them, because they won't be injected when the first mod assembly fails to compile.  😛

Like you, I wanna see mod-management get some TLC, but I'm afraid it's not us modders who can really give it the attention it needs.  ...not directly, anyways.  We're dependent on the game's developers to make improvements like these for us.
